modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,243 @@
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
+                   libMesh::Utility::enum_to_string(curr_edge->type()),
+                   " base element type is: ",
+                   libMesh::Utility::enum_to_string(elem->type()));
+      const Node * node1 = curr_edge->node_ptr(0);
+      const Node * node2 = curr_edge->node_ptr(1);
+
+      for (const auto & cut_elem : _cutter_mesh->element_ptr_range())
+      {
+        std::vector<Point> vertices;
+
+        for (auto & node : cut_elem->node_ref_range())
+        {
+          Point & this_point = node;
+          vertices.push_back(this_point);
+        }
+
+        Point intersection;
+        if (Xfem::intersectWithEdge(*node1, *node2, vertices, intersection) &&
+            std::find(cut_edges.begin(), cut_edges.end(), j) == cut_edges.end())
+        {
+          cut_edges.push_back(j);
+          cut_pos.emplace_back(Xfem::getRelativePosition(*node1, *node2, intersection));
+        }
+      }
+    }
+
+    // if two edges of an element are cut, it is considered as an element being cut
+    if (cut_edges.size() == 2)
+    {
+      elem_cut = true;
+      Xfem::CutFace mycut;
+      mycut._face_id = i;
+      mycut._face_edge.push_back(cut_edges[0]);
+      mycut._face_edge.push_back(cut_edges[1]);
+      mycut._position.push_back(cut_pos[0]);
+      mycut._position.push_back(cut_pos[1]);
+      cut_faces.push_back(mycut);
+    }
+  }
+
+  return elem_cut;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_edges*/,
+    std::vector<Xfem::CutEdge> & /*cut_edges*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_faces*/,
+    std::vector<Xfem::CutFace> & /*cut_faces*/) const
+{
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+Real
+InterfaceMeshCut3DUserObject::calculateSignedDistance(Point p) const
+{
+  std::vector<Real> distance;
+  Real min_dist = std::numeric_limits<Real>::max();
+  for (const auto & cut_elem : _cutter_mesh->element_ptr_range())
+  {
+    std::vector<Point> vertices{
+        cut_elem->node_ref(0), cut_elem->node_ref(1), cut_elem->node_ref(2)};
+    unsigned int region;
+    Point xp;
+    Real dist = Xfem::pointTriangleDistance(
+        p, cut_elem->node_ref(0), cut_elem->node_ref(1), cut_elem->node_ref(2), xp, region);
+
+    distance.push_back(std::abs(dist));
+
+    if (dist < std::abs(min_dist))
+    {
+      min_dist = dist;
+      Point normal = (_pseudo_normal.find(cut_elem->id())->second)[region];
+      if (normal * (p - xp) < 0.0)
+        min_dist *= -1.0;
+    }
+  }
+  std::sort(distance.begin(), distance.end());
+  Real sum_dist = 0.0;
+  for (std::vector<Real>::iterator it = distance.begin(); it != distance.begin() + 1; ++it)
```

Comment:
This is pretty different from the behavior of the 2D equivalent. The 2D version just returns min_dist regardless of whether it's positive or negative. Is there a reason they're not consistent?

Reply:
For 3D, I used the sum of first two minimum distances as the final distance. If I remember correctly, it helps smoothing the signed distance values.